# Import libraries

In [59]:
import pandas as pd
import numpy as np
import os

## Import Dataframes

In [60]:
path = path = r"C:\Users\Poory\OneDrive\Desktop\project\Instacart Basket Analysis"

In [61]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [62]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [63]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column              Dtype  
---  ------              -----  
 0   Unnamed: 0          int64  
 1   order_id            int64  
 2   user_id             int64  
 3   order_number        int64  
 4   orders_day_of_week  int64  
 5   order_hour_of_day   int64  
 6   last_order_days     float64
dtypes: float64(1), int64(6)
memory usage: 182.7 MB


In [64]:
# I decided to change the user_id and Order_id types to string again
df_ords['order_id'] = df_ords['order_id'].astype('str')
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [65]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column              Dtype  
---  ------              -----  
 0   Unnamed: 0          int64  
 1   order_id            object 
 2   user_id             object 
 3   order_number        int64  
 4   orders_day_of_week  int64  
 5   order_hour_of_day   int64  
 6   last_order_days     float64
dtypes: float64(1), int64(4), object(2)
memory usage: 182.7+ MB


#### Mixed-Type Data

In [66]:
# create dataframe
df_test = pd.DataFrame()

In [67]:
# creat a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [68]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


C:\Users\Poory\AppData\Local\Temp\ipykernel_9908\772068622.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [69]:
# Check for columns with inconsistent data types in orders (chatGPT version)
for col in df_ords.columns.tolist():
    
    # Check if there are any inconsistencies in data types within the column
    weird = (df_ords[[col]].apply(lambda x: x.map(type)) != df_ords[[col]].iloc[0].apply(type)).any(axis=1)
    
    # If inconsistencies are found, print the column name
    if len(df_ords[weird]) > 0:
        print(col)


In [70]:
# Check for columns with inconsistent data types in products
for col in df_prods.columns.tolist():
    weird = (df_prods[[col]].apply(lambda x: x.map(type)) != df_prods[[col]].iloc[0].apply(type)).any(axis=1)
    if len(df_prods[weird]) > 0:
        print(col)

product_name


In [71]:
df_prods['product_name'].dtype

dtype('O')

#### Missing Values:

##### Finding Missing Values

In [72]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [73]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14   12.20
68             69          NaN        26              7   11.80
115           116          NaN        93              3   10.80
261           262          NaN       110             13   12.10
525           525          NaN       109             11    1.20
1511         1511          NaN        84             16   14.30
1780         1780          NaN       126             11   12.30
2240         2240          NaN        52              1   14.20
2586         2586          NaN       104             13   12.40
3159         3159          NaN       126             11   13.10
3230         3230          NaN       120             16   14.40
3736         3736          NaN        41              8   14.80
4283         4283          NaN        77              7   14.40
4790         4790          NaN        91             16   14.50
38187       38183          NaN        39             12   20.90
40444       40440          NaN       120             16   14.80

##### Addressing Missing Values

##### Impute missing values by Mean or Median

In [74]:
# df['column with missings'].fillna(mean value, inplace=True)
# Finding Mean: df.describe()

In [75]:
# df['column with missings'].fillna(median value, inplace=True)
# Finding Median: df.median()

In [76]:
df_prods.shape

(49693, 5)

In [77]:
df_prods2 = df_prods[df_prods['product_name'].isnull() == False]
df_prods2.shape

(49677, 5)

In [78]:
# Dropping all missing values:
# df.dropna(inplace = True)
# Dropping all missing values in specific column:
# df.dropna(subset = [‘Column name’], inplace = True) * overwrite main df *

#### Duplicates:

In [79]:
# Looking for full duplicates
df_dups = df_prods2[df_prods2.duplicated()]
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11    4.80  
18459        27              5    9.20  
26810        27              5   13.40  
35309       121             14    6.80  
35495       127             11    9.90

In [80]:
# Deleting duplicated values
df_prods3 = df_prods2.drop_duplicates()
df_prods3.shape

(49672, 5)

In [81]:
# Finding Mixed type columns in prods3 df:
for col in df_prods3.columns.tolist():
    weird = (df_prods3[[col]].apply(lambda x: x.map(type)) != df_prods3[[col]].iloc[0].apply(type)).any(axis = 1)
    if len(df_prods3[weird]) > 0:
        print(col)

In [82]:
# Final products df 
df_prods3.shape

(49672, 5)

### Export Products DF

In [83]:
df_prods3.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'Products_checked.csv'))

# Task 4.5 : Consistency check on Orders df

In [84]:
#Q2:
df_ords.describe()

Unnamed: 0  order_number  orders_day_of_week  order_hour_of_day  \
count  3421083.00    3421083.00          3421083.00         3421083.00   
mean   1710541.00         17.15                2.78              13.45   
std     987581.74         17.73                2.05               4.23   
min          0.00          1.00                0.00               0.00   
25%     855270.50          5.00                1.00              10.00   
50%    1710541.00         11.00                3.00              13.00   
75%    2565811.50         23.00                5.00              16.00   
max    3421082.00        100.00                6.00              23.00   

       last_order_days  
count       3214874.00  
mean             11.11  
std               9.21  
min               0.00  
25%               4.00  
50%               7.00  
75%              15.00  
max              30.00

In [85]:
# Change display format to avoid scientific notation (chatGPT)
pd.set_option('display.float_format', lambda x: '%.2f' % x)
df_ords.describe()

Unnamed: 0  order_number  orders_day_of_week  order_hour_of_day  \
count  3421083.00    3421083.00          3421083.00         3421083.00   
mean   1710541.00         17.15                2.78              13.45   
std     987581.74         17.73                2.05               4.23   
min          0.00          1.00                0.00               0.00   
25%     855270.50          5.00                1.00              10.00   
50%    1710541.00         11.00                3.00              13.00   
75%    2565811.50         23.00                5.00              16.00   
max    3421082.00        100.00                6.00              23.00   

       last_order_days  
count       3214874.00  
mean             11.11  
std               9.21  
min               0.00  
25%               4.00  
50%               7.00  
75%              15.00  
max              30.00

###### It seems everything is good. days of the week and hours of a day are shown correctly and there isn't any negative number in the last order days, but we can see there are about 200000 missing values in last order days column.

In [86]:
#Q3 Check for mixed-type data in your df_ords dataframe:
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].apply(lambda x: x.map(type)) != df_ords[[col]].iloc[0].apply(type)).any(axis=1)
    if len(df_ords[weird]) > 0:
        print(col)

###### It seems there isn't any mixed type column in orders data frames 

In [87]:
#Q5 Finding missing values
df_ords.isnull().sum()

Unnamed: 0                 0
order_id                   0
user_id                    0
order_number               0
orders_day_of_week         0
order_hour_of_day          0
last_order_days       206209
dtype: int64

###### There are 206209 missing values in the last_order_days column. I think these are our new customers who had an order for the first time, so there isn't any value for these customers in this column. In addition, these missing values could be our customers' first order.

In [88]:
df_nan2 = df_ords[df_ords['last_order_days'].isnull() == True]
df_nan2

Unnamed: 0 order_id user_id  order_number  orders_day_of_week  \
0                 0  2539329       1             1                   2   
11               11  2168274       2             1                   2   
26               26  1374495       3             1                   1   
39               39  3343014       4             1                   6   
45               45  2717275       5             1                   3   
...             ...      ...     ...           ...                 ...   
3420930     3420930   969311  206205             1                   4   
3420934     3420934  3189322  206206             1                   3   
3421002     3421002  2166133  206207             1                   6   
3421019     3421019  2227043  206208             1                   1   
3421069     3421069  3154581  206209             1                   3   

         order_hour_of_day  last_order_days  
0                        8              NaN  
11                      11              NaN  
26                      14              NaN  
39                      11              NaN  
45                      12              NaN  
...                    ...              ...  
3420930                 12              NaN  
3420934                 18              NaN  
3421002                 19              NaN  
3421019                 15              NaN  
3421069                 11              NaN  

[206209 rows x 7 columns]

In [89]:
df_nan2['order_number'].value_counts(dropna = False)

order_number
1    206209
Name: count, dtype: int64

### As we can see all of our number 1 orders have NaN values in last_order_days. So, i want to make new column and devide all orders in two group: first order and second and more orders.

In [90]:
df_ords['first_order'] = False
df_ords.loc[df_ords['order_number'] == 1, 'first_order'] = True
df_ords

Unnamed: 0 order_id user_id  order_number  orders_day_of_week  \
0                 0  2539329       1             1                   2   
1                 1  2398795       1             2                   3   
2                 2   473747       1             3                   3   
3                 3  2254736       1             4                   4   
4                 4   431534       1             5                   4   
...             ...      ...     ...           ...                 ...   
3421078     3421078  2266710  206209            10                   5   
3421079     3421079  1854736  206209            11                   4   
3421080     3421080   626363  206209            12                   1   
3421081     3421081  2977660  206209            13                   1   
3421082     3421082   272231  206209            14                   6   

         order_hour_of_day  last_order_days  first_order  
0                        8              NaN         True  
1                        7            15.00        False  
2                       12            21.00        False  
3                        7            29.00        False  
4                       15            28.00        False  
...                    ...              ...          ...  
3421078                 18            29.00        False  
3421079                 10            30.00        False  
3421080                 12            18.00        False  
3421081                 12             7.00        False  
3421082                 14            30.00        False  

[3421083 rows x 8 columns]

### I want to address the issue of missing values in our customer data. Simply imputing the mean or median values wouldn't be appropriate, as these missing values are related to our customers' first orders. However, deleting this data could significantly distort our overall data frame. To ensure we don't lose any valuable information, I  added a new column that divides all orders into two groups: first orders and second/more orders. This approach should help us better manage the missing data and ensure that our analysis is accurate.

In [91]:
#Q7 & Q8 Finding and Addressing duplicated values
df_dup2 = df_ords[df_ords.duplicated()]
df_dup2

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, last_order_days, first_order]
Index: []

##### It seems we don't have any duplicated value.

In [92]:
df_ords = df_ords.drop_duplicates()
df_ords.shape

(3421083, 8)

# Exporting Orders df

In [93]:
df_ords2.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))